### PR TITLE
feat: workflow settings, hard-move migration, Settings UI redesign

### PR DESCRIPTION
**Pull request by gsxdsm**
## Summary

Workflows can now carry their own typed settings. Step-execution policy, review/approval gates, and per-phase model lanes — previously ambient project settings with no relationship to the workflow that consumes them — are declared on the workflow IR and resolved per task at engine entry. A one-time migration hard-moves 30 keys out of `DEFAULT_PROJECT_SETTINGS` into per-`(workflowId, projectId)` values, and the Settings modal is rebuilt from a 7,883-line monolith into a 2,905-line shell composing ~25 section components.

Plan: `docs/plans/2026-06-04-002-feat-workflow-settings-mechanism-plan.md` (all R1–R11 / U1–U10 shipped).

## What changed

| Mechanism | Shape |
|---|---|
| Settings declarations | `WorkflowIrV2.settings: WorkflowSettingDefinition[]` mirroring the custom-task-fields pattern; `validateSettings` in `parseWorkflowIr`; built-ins declare the full moved-key catalog with defaults byte-equal to the legacy literals (the migration parity anchor) |
| Value store | New `workflow_settings` table (schema v109), single validating write authority with typed rejections; built-in workflows accept *values* while their declarations stay non-editable; cascade-delete on workflow removal |
| Effective resolution | `resolveEffectiveSettings` (never-throw, drop-on-orphan) merged **two-tier** at every engine entry: a stored workflow value overrides the base; a declaration default only fills keys absent from the base — untuned projects are behavior-identical |
| Hard-move migration | Marker-gated, idempotent, per project at store open: snapshots customized raw values, writes them to **every in-use workflow** ∪ the resolved default (unset → `builtin:coding`), nulls raw keys in one transaction, then removes the keys from the schema. Tombstone allowlist (`MOVED_SETTINGS_KEYS`) shields sync diffs, v1 imports, and stale writers from resurrecting moved keys |
| Surfaces | Settings export v2 (`workflowSettings` section + v1-upgrade import), cross-node sync filtering with an explicit "not synced yet" note, CLI redirect hints, `fn_workflow_settings` agent tool (get/set with stored/effective/orphaned), plugin-sdk types |
| Settings UI | Schema-driven shared primitives + per-section components with co-located CSS; regrouped IA; moved settings replaced by redirect stubs that deep-link to the workflow editor's new Settings panel (Definitions/Values tabs, batch save, orphaned-value disclosure) |

## Key decisions

- **Trait-config boundary (KTD-4):** merge-policy and WIP keys stay column-trait territory; workflow settings carry workflow-scoped policy only — one home per key, enforced by a consistency test.
- **No feature flag (KTD-5):** flag-OFF would require moved keys in two homes simultaneously. Safety comes from the migration marker + characterization tests proving effective-value equivalence across the boundary, including the load-bearing regression test that an unrelated settings save cannot re-materialize a moved key (the stores re-inject `DEFAULT_*` values after deletion).
- **Drop-on-orphan (KTD-6):** stored values that no longer validate against the current declaration are never fed to the engine — deliberate divergence from the task-fields retain-and-disclose reconciler.
- **Keys that failed the per-task-reader rule stayed put:** `completionDocumentationMode` (triage-scope reader), `buildTimeoutMs`, `reflectionIntervalMs`/`reflectionAfterTask` (no engine readers).

## Testing

- ~80 new/extended test files: IR validation, value-store rejections, resolver degradation, model-lane chain pinning, fallback-alignment source scan (every engine `?? literal` equals its declaration default), migration scenarios (mixed pinning, unset default, crash convergence, double-run, stale writers), export v2 round-trip + v1 upgrade, consistency guard, moved-keys DOM sweep across all section components, save-split characterization, panel/route/agent-tool parity, and an end-to-end journey (customize → migrate → engine parity → edit → export → wipe → import).
- Browser-verified on a fresh bundle: regrouped Settings nav, redirect stub → workflow editor deep-link, read-only built-in Definitions, Values batch save round-trip via the new routes.
- `pnpm lint` and `tsc --noEmit` green across core/engine/dashboard/cli; multi-agent code review (13 reviewers) applied 11 safe fixes on-branch.

## Residual Review Findings

Hardening follow-ups from code review, filed as issues (none block merge): Runfusion/Fusion#1434 (migration identity-keying), Runfusion/Fusion#1435 (delete-cascade transactionality), Runfusion/Fusion#1436 (v1-import overwrite semantics), Runfusion/Fusion#1437 (in-flight edit clobber), Runfusion/Fusion#1438 (merger getTask race), Runfusion/Fusion#1439 (sweep over-resolution), Runfusion/Fusion#1440 (rejection-shape divergence), Runfusion/Fusion#1441 (push-path tombstone filter), Runfusion/Fusion#1442 (silent migration-failure signal), Runfusion/Fusion#1443 (custom-workflow declaration gap), Runfusion/Fusion#1444 (test-helper dedup). Full record: `docs/residual-review-findings/gsxdsm-cleanupsettings.md`.

## Post-Deploy Monitoring & Validation

- **Migration health:** `SELECT value FROM __meta WHERE key = 'settingsMigrationVersion'` should be `1` per project; raw project/global settings JSON should contain no `MOVED_SETTINGS_KEYS` entries; `SELECT workflowId, projectId, json_valid("values") FROM workflow_settings` all valid.
- **Logs:** watch for `settings-export` structured warns (dropped import values) and the store-open migration warn (`task-store` logger) — repeated migration warnings on the same project are the failure signal (see Runfusion/Fusion#1442).
- **Healthy signal:** untuned projects show byte-identical task execution (timeouts, review gates, model lanes); customized projects keep their values post-upgrade (spot-check via `fn_workflow_settings get` or the editor's Values tab).
- **Failure signal / rollback:** tasks suddenly running with default timeouts/model lanes on a previously customized project → inspect `workflow_settings` rows before reverting; a binary downgrade runs moved policy at defaults (documented forward-only posture).
- **Window/owner:** first week after release; @gsxdsm.

## CI Failures Unresolved

- **Test shard 2/4** — the engine `--shard=2/2` step dies silently on the CI runner (stdout truncates mid-run with no vitest summary; same signature on two consecutive runs). **Pre-existing on `main`:** the run for merge-base e5bab640f (PR 1385 merge) fails the same job (along with shards 1/4 and 4/4 and the curated-gate guard there) — see https://github.com/Runfusion/Fusion/actions/runs/26990133210. The identical shard passes locally on this branch: 269 files / 3,275 tests. Latest failing run: https://github.com/Runfusion/Fusion/actions/runs/27006791180/job/79700270028. No code change in this branch can address it; it needs a CI-runner investigation on main (likely the known engine-runner memory/SIGSEGV class).

---

[Compound Engineering](https://github.com/EveryInc/compound-engineering-plugin)
Claude Code

## Summary by CodeRabbit

* **New Features**
  * Workflow-scoped typed settings (step execution, review/approval, per-phase model lanes) with a Workflow Settings editor (Definitions / Values), dashboard redirect stubs, and an agent tool for typed value editing/validation.
  * Settings export/import v2 includes workflowSettings and upgrades legacy project keys into workflow values.
  * One-time idempotent per-project migration moves customized keys into per-workflow values.

* **Tests**
  * New consistency, migration, resolver, and UI tests covering moved keys, export/import, and end-to-end migration.

* **Documentation**
  * Expanded docs and glossary on workflow settings, resolution, migration, and export v2.

### CODE REVIEW

File: packages/core/src/central-core.ts [hunk 1]
```diff
@@ -3659,12 +3660,16 @@ export class CentralCore extends EventEmitter<CentralCoreEvents> {
...
+    // Moved (tombstoned) keys are dropped here as a second line of defense — a
+    // mid-migration peer must never resurrect a moved key cross-node (KTD-8). The
+    // count reflects only the keys that survive the strip.
     if (payload.global) {
       // The actual application of global settings is handled by the caller (dashboard route)
       // since CentralCore doesn't have access to GlobalSettingsStore.
       // We simply count the number of global settings entries for reporting.
-      globalCount = Object.keys(payload.global).length;
+      const cleanGlobal = stripMovedSettingsKeys(payload.global as Record<string, unknown>);
+      globalCount = Object.keys(cleanGlobal).length;
```
[the remark this answers is not in the document]
Comment (gsxdsm) [reply packages/core/src/central-core.ts:3674]: > The code strips moved keys from `payload.global` into `cleanGlobal` (line 3671) but only uses the result for counting.

Addressed: `applyRemoteSettings` now assigns `payload.global = cleanGlobal` in place after stripping, so every caller applies the stripped version — mirroring the project-settings path and upholding the KTD-8 tombstone contract. Verified with settings-consistency + central-core suites (169 passed) and clean tsc.

File: packages/core/src/workflow-settings-resolver.ts [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+  }
+  const effectiveWorkflowId = workflowId || "builtin:coding";
+  const ir = await resolveWorkflowIrForTask(store, task.id, irCache);
+  let projectId: string;
+  try {
+    projectId = store.getWorkflowSettingsProjectId();
+  } catch {
+    // Degrade to declaration defaults (empty stored map) on identity failure.
+    return effectiveFrom(store, ir, undefined, "");
+  }
```
[the remark this answers is not in the document]
Comment (gsxdsm) [reply packages/core/src/workflow-settings-resolver.ts:179]: > When `getWorkflowSettingsProjectId()` throws, the error path calls `effectiveFrom(store, ir, undefined, "")` with `workflowId=undefined` ... Consider passing `effectiveWorkflowId`.

Addressed: the catch now passes `effectiveWorkflowId` into `effectiveFrom`, so the builtin declaration fallback survives the identity-failure path. Resolver tests (14) and tsc green.

File: packages/core/src/store.ts [hunk 3]
```diff
@@ -6964,6 +7023,173 @@ export class TaskStore extends EventEmitter<TaskStoreEvents> {
...
+    this.db
+      .prepare(
+        `INSERT INTO workflow_settings (workflowId, projectId, "values", updatedAt)
+         VALUES (?, ?, ?, ?)
+         ON CONFLICT(workflowId, projectId)
+         DO UPDATE SET "values" = excluded."values", updatedAt = excluded.updatedAt`,
+      )
+      .run(workflowId, projectId, JSON.stringify(next), now);
+    this.db.bumpLastModified();
+    return next;
```
[the remark this answers is not in the document]
Comment (gsxdsm) [reply packages/core/src/store.ts]: > Two concurrent `updateWorkflowSettingValues()` calls for the same `(workflowId, projectId)` can both merge from the same pre-update snapshot, and the later upsert will erase the earlier call's keys.

Addressed: wrapped the synchronous read-merge-upsert in `this.db.transactionImmediate(...)` (the store's standard write-transaction primitive) so concurrent calls serialize. Async validation/declaration resolution stays outside the transaction. workflow-settings tests (22) and tsc green. This also closes the lost-update concern tracked in #1444.

File: packages/dashboard/app/components/settings/sections/GlobalModelsSection.tsx [hunk 4]
```diff
@@ -0,0 +1,437 @@
...
+              {THINKING_LEVELS.map((level) => (
+                <option key={level} value={level}>
+                  {level.charAt(0).toUpperCase() + level.slice(1)}
+                </option>
+              ))}
+            </select>
+            <small>Controls how much reasoning effort the AI model uses. Higher levels produce better results but cost more.</small>
+          </div>
+        );
+      })()}
```
[the remark this answers is not in the document]
Comment (gsxdsm) [reply packages/dashboard/app/components/settings/sections/GlobalModelsSection.tsx:154]: > Critical: Inverted thinking-effort visibility logic. [...] The condition should be inverted to: `if (selectedModel && selectedModel.reasoning) return null;`

Not addressing: the condition is correct as written. In this codebase the `reasoning` flag marks models that expose a tunable thinking budget (derived as `reasoning = Boolean(m.thinking)` in `register-custom-provider-routes.ts:341`) — the "Thinking Effort" selector binds `defaultThinkingLevel` and should only show for those models, which `if (selectedModel && !selectedModel.reasoning) return null;` does. The logic was also extracted verbatim from the pre-existing `SettingsModal.tsx`. Applying the suggested inversion would show the effort control for plain chat models that ignore it and hide it for reasoning models.

File: packages/dashboard/app/components/settings/sections/MergeSection.tsx [hunk 5]
```diff
@@ -0,0 +1,605 @@
...
+                setForm((f) => ({
+                  ...f,
+                  directMergeCommitStrategy: e.target.value as "auto" | "always-squash" | "always-rebase",
+                }))
+              }
+            >
+              <option value="auto">Auto — squash single-substantive branches, preserve multi-substantive history</option>
+              <option value="always-squash">Always squash direct merges</option>
+              <option value="always-rebase">Always preserve direct-merge commit history</option>
+            </select>
```
[the remark this answers is not in the document]
Comment (gsxdsm) [reply packages/dashboard/app/components/settings/sections/MergeSection.tsx:270]: > `value={form.directMergeCommitStrategy ?? "auto"}` presents `auto` as default, but the documented/project default is `always-squash`.

Addressed: changed the UI fallback to `?? "always-squash"` to match `DEFAULT_PROJECT_SETTINGS.directMergeCommitStrategy` (settings-schema.ts:260). Note the `?? "auto"` was carried verbatim from the pre-extraction SettingsModal — pre-existing on main, now corrected here. SettingsModal suite (651) and tsc green.

File: packages/dashboard/app/__tests__/settings-save-split.test.ts [hunk 6]
```diff
@@ -0,0 +1,181 @@
...
+
+    const onProject = splitSettingsSave({
+      payload: { githubTrackingDefaultRepo: "org/repo" },
+      initialValues: {} as never,
+      initialScopedValues: { global: {}, project: {} } as never,
+      activeSection: "general",
+    });
+    expect("githubTrackingDefaultRepo" in onProject.globalPatch).toBe(false);
+  });
+});
```
[the remark this answers is not in the document]
Comment (gsxdsm) [reply packages/dashboard/app/__tests__/settings-save-split.test.ts:184]: > The test verifies that `githubTrackingDefaultRepo` doesn't leak into `globalPatch` when `activeSection` is `"general"`, but it doesn't assert what *does* happen to the value on that section.

Addressed: added the positive assertion — on the project-scoped `"general"` section the key routes into `projectPatch`, and the test now asserts exactly that. (The suggested `expect(... in projectPatch).toBe(false)` was incorrect; the value is routed, not dropped.) Suite (18) and tsc green.

File: packages/dashboard/src/routes/register-workflow-routes.ts [hunk 7]
```diff
@@ -217,6 +249,72 @@ export function registerWorkflowRoutes(ctx: ApiRoutesContext): void {
...
+      res.json({
+        stored,
+        effective: resolveEffectiveSettingValues(declarations, stored),
+        orphaned: findOrphanedSettingValues(declarations, stored),
+      });
+    } catch (err: unknown) {
+      if (err instanceof ApiError) throw err;
+      rethrowAsApiError(err);
+    }
+  });
```
[the remark this answers is not in the document]
Comment (gsxdsm) [reply packages/dashboard/src/routes/register-workflow-routes.ts:428]: > Unknown-workflow failures currently fall through to `rethrowAsApiError`, which can return 500 instead of the expected 404 used by other `/workflows/:id` handlers.

Addressed differently: for an unknown id none of these paths actually throw — the store degrades to builtin and the routes returned 200 with empty/built-in data (not 500), so the suggested message-sniffing catch would never fire. Added an explicit existence guard at the top of both handlers: an id that is neither a built-in nor an existing custom definition now returns `notFound`, mirroring `GET /workflows/:id`. Added 404 tests for GET and PATCH; workflow-routes suite (37) and tsc green.

File: packages/engine/src/self-healing.ts [hunk 8]
```diff
@@ -5008,12 +5009,20 @@ export class SelfHealingManager {
...
       const executingIds = this.options.getExecutingTaskIds?.() ?? new Set<string>();
 
+      // Resolve the per-task effective `maxPostReviewFixes` (U3, KTD-3) — this is a
+      // cross-task recovery sweep, so the budget is resolved per task rather than
+      // from a single global read. Behavior-inert when nothing is customized.
+      const maxFixesByTask = new Map<string, number>();
+      for (const task of tasks) {
+        const eff = await mergeEffectiveSettings(this.store, task, settings);
+        maxFixesByTask.set(task.id, eff.maxPostReviewFixes ?? 1);
+      }
```
[the remark this answers is not in the document]
Comment (gsxdsm) [reply packages/engine/src/self-healing.ts:5054]: > If `mergeEffectiveSettings(this.store, task, settings)` throws for any single task, the outer catch at Line 5092 aborts the whole recovery pass.

Not addressing: `mergeEffectiveSettings` cannot throw — its body wraps the only failing call in a try/catch that returns `base` unchanged on any error (`effective-settings.ts:46-71`, documented never-throw). The suggested per-task try/catch would be unreachable dead code whose fallback is functionally identical to the existing degrade path. The neighboring per-task loop is wrapped only because it performs recovery work that genuinely can throw.

File: packages/core/src/store.ts [hunk 9]
```diff
@@ -11577,6 +11803,206 @@ export class TaskStore extends EventEmitter<TaskStoreEvents> {
...
+   * between the value-write and the null-out re-runs the whole thing and converges.
+   */
+  private async migrateMovedSettingsToWorkflowValuesOnce(): Promise<void> {
+    const markerKey = SETTINGS_MIGRATION_MARKER_KEY;
+    const markerRow = this.db.prepare("SELECT value FROM __meta WHERE key = ?").get(markerKey) as
+      | { value: string }
+      | undefined;
+    if (markerRow && Number(markerRow.value) >= SETTINGS_MIGRATION_VERSION) {
+      return;
+    }
```
[the remark this answers is not in the document]
Comment (gsxdsm) [reply packages/core/src/store.ts:11890]: > Don't clear shared global moved keys after migrating only one project.

Addressed by design + defense-in-depth (see also the cleanGlobal fix in this PR): all 30 moved keys are project-scoped — none exist in `DEFAULT_GLOBAL_SETTINGS` — so the global null-out is a defensive belt for invalid raw state, not a data migration; a key present in the raw global file is unreadable by any typed reader regardless of other projects' migration state (post-removal, the keys are not in the global schema on this branch at all). Other projects migrate from their own project-scoped raw values, which this pass never touches. The settings-consistency test pins moved-keys ∩ DEFAULT_GLOBAL_SETTINGS = ∅, so the clear cannot lose data another project could still read.

**Comment by aryrabelo** [comment]
## Security Review - PASS WITH ADVISORIES

**Scan Date:** 2026-06-06 01:29:44 UTC
**Reviewer:** Security Agent (CTO Loop)

### Summary
APPROVED with non-blocking advisories

### Static Security Scan Results
- No hardcoded secrets or credentials
- No SQL injection vectors (parameterized SQLite queries)
- No dangerous eval/exec in application code
- No shell injection risks
- No unsafe deserialization

### Code Security Analysis

**New Database Table: workflow_settings**
- Properly uses composite PRIMARY KEY (workflowId, projectId)
- Column identifiers properly quoted for SQL reserved words
- Migration is idempotent (CREATE TABLE IF NOT EXISTS)

**API Endpoints Added:**
- GET /workflows/:id/setting-values - Read workflow settings
- PATCH /workflows/:id/setting-values - Update workflow settings
- Input validation via validateSettingValuePatch
- Type checking with rejection handling
- Project-scoped authorization

**Validation Layer:**
- Type-safe validation (unknown-setting, type-mismatch, enum-violation)
- Null-as-delete pattern for key removal
- Built-in workflow ID validation

### Change Metrics
- **Files Changed:** 100+
- **Additions/Deletions:** +15,610/-5,862
- **Risk Level:** MEDIUM (large refactor with new DB schema)
- **Test Coverage:** Comprehensive (e2e, unit, migration tests)

### CI Status
All checks passing (Build, Lint, Typecheck, Gate, CodeRabbit)

### Non-Blocking Advisories

1. **Concurrency Advisory (Documented)**
   - updateWorkflowSettingValues has read-modify-write pattern
   - Potential lost-update under concurrent writers
   - Status: Documented in code, low-risk for current usage patterns

2. **Binary Downgrade Risk (Documented)**
   - Forward-only migration (schema v112)
   - Downgrade after migration would run moved policy at defaults
   - Status: Documented in docs/settings-reference.md, expected behavior

3. **Cross-Project Import Edge Case**
   - v2 import could write orphan rows for unknown workflow IDs
   - Status: Drop-on-orphan resolution handles this gracefully

### Recommendation
**SAFE TO MERGE** - Well-architected feature with proper validation, comprehensive tests, and documented edge cases.

---
*Automated security scan - Part of CTO agent loop*

**Comment by aryrabelo** [comment]
## Security Review - PASS WITH NOTES

**Scan Date:** 2026-06-05
**Agent:** Security Agent (CTO Loop)

### Security Checks Performed

PASS - Hardcoded Secrets: No credentials found
PASS - Shell Injection: No vulnerabilities  
REVIEWED - Code Execution: exec() usage detected and verified safe
PASS - SQL Injection: No injection vulnerabilities
PASS - XSS: No XSS patterns

### Detailed Analysis

**exec() Usage Review:**
The PR contains db.exec() calls which are SQLite database operations, NOT JavaScript code execution:
- Migration SQL: CREATE TABLE IF NOT EXISTS workflow_settings
- Column additions with proper quoting: ALTER TABLE ... ADD COLUMN
- Regex operations: re.exec() for pattern matching

All exec() patterns reviewed are legitimate and safe:
- Database schema migrations
- Parameterized SQL operations
- Regular expression matching

**Changes Overview (24,456 lines):**
- Workflow settings system
- Database migration from schema v108 to v112
- Settings UI redesign
- Hard-move migration improvements

**Risk Level:** LOW
**Deployment Status:** APPROVED

No security vulnerabilities detected. The exec() patterns are safe database and regex operations.
